src/Workspaces/Core/Portable/Rename/Renamer.RenameSymbolDocumentAction.cs

```diff
@@ -0,0 +1,115 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+#nullable enable
+
+using System;
+using System.Collections.Immutable;
+using System.Globalization;
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.LanguageServices;
+using Microsoft.CodeAnalysis.Options;
+using Microsoft.CodeAnalysis.Shared.Extensions;
+
+namespace Microsoft.CodeAnalysis.Rename
+{
+    public static partial class Renamer
+    {
+        /// <summary>
+        /// Action that will rename a type to match the current document name
+        /// </summary>
+        internal class RenameSymbolDocumentAction : RenameDocumentAction
+        {
+            private readonly AnalysisResult _analysis;
+
+            private RenameSymbolDocumentAction(
+                AnalysisResult analysis,
+                OptionSet optionSet,
+                ImmutableArray<ErrorResource> errors)
+                : base(errors, optionSet)
+            {
+                _analysis = analysis;
+            }
+
+            public override string GetDescription(CultureInfo? culture)
+                => string.Format(WorkspacesResources.ResourceManager.GetString("Rename_0_to_1", culture ?? WorkspacesResources.Culture)!, _analysis.OriginalDocumentName, _analysis.NewDocumentName);
+
+            internal override async Task<Solution> GetModifiedSolutionAsync(Document document, CancellationToken cancellationToken)
+            {
+                var solution = document.Project.Solution;
+                var matchingTypeDeclaration = await GetMatchingTypeDeclarationAsync(document, _analysis.OriginalSymbolName!, cancellationToken).ConfigureAwait(false);
+
+                if (matchingTypeDeclaration is object)
+                {
+                    var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+                    var symbol = semanticModel.GetDeclaredSymbol(matchingTypeDeclaration, cancellationToken);
+
+                    solution = await RenameSymbolAsync(solution, symbol, _analysis.NewSymbolName, OptionSet, cancellationToken).ConfigureAwait(false);
+                }
+
+                return solution;
```

Comment:
this feels strange.  why would we have the rename action if it could fail to modify the solution?

Reply:
Because we are updating some solution that may or may not be the actual analyzed solution. If for some reason the user of the API changed the document contents before calling update then we should fail gracefully.

Reply:
so i would *much* rather have us fail *hard* in that case.  i.e. that indicates something super weird on the callers side.  i.e. they got this info, then just applied these options ot a random *different* solution.

Can we not throw in that case?

in fact, can we just have the rename-actions just hold onto the Doc they applied to.  then, if you try to apply the action to a different solution than the one in the doc, we just fail outright (with a good diagnostic message).

Reply:
ok.  so my comment from 3 weeks ago still applies.   :)

Reply:
@jasonmalinowski and I had a discussion about this. It's too hard for the caller to make sure the correct solution/document is used when applying the actions. Especially in the case of the project system which is the main consumer.

Reply:
can we have a chat?  i guess i'm not understanding why that would be hard.  i would expect:

1. the PS will call into us to ask us for the set of rename actions.
2. they prompt the user to see what the user wants to apply
3. they apply the chosen options.

It shouldn't be able to fail.  Note: this is how code-actions work.  We get queried for the set of actions, the user is shown the actions, they can choose one.  The action applies to the solution that was used to compute it.

Reply:
I'll set up something tomorrow